libs/hmpb/render-backend/src/next/render_ballot.tsx

```diff
@@ -0,0 +1,233 @@
+import { Buffer } from 'buffer';
+import { assertDefined } from '@votingworks/basics';
+import {
+  PdfOptions,
+  PixelDimensions,
+  createRenderer,
+  RenderDocument,
+} from './renderer';
+
+export const contentSlot = (
+  <div id="content-slot" style={{ height: '100%', width: '100%' }} />
+);
+
+export const qrCodeSlot = <div data-type="QrCodeSlot" />;
+
+export type FrameComponent<P> = (
+  props: P & { children: JSX.Element; pageNumber: number; totalPages: number }
+) => JSX.Element;
+
+export interface PagedElementResult<P> {
+  currentPageElement: JSX.Element;
+  nextPageProps?: P;
+}
+
+export type ContentComponent<P> = (
+  props: P & { dimensions: PixelDimensions },
+  // The content component is passed the document so that it can measure
+  // elements in order to determine how much content fits on each page.
+  document: RenderDocument
+) => Promise<PagedElementResult<P>>;
+
+/**
+ * A page template consists of two interlocking pieces:
+ * - A frame component (imagine it like a picture frame) that is rendered on each page
+ * - A content component that knows how to render a page at a time of content
+ * within the frame. Given a set of props (e.g. list of contests) the content component returns two items:
+ *     - The content element for the current page (e.g. the contest boxes for this page)
+ *     - The props for the next page (e.g. the contests that didn't fit on this page)
+ */
+export interface BallotPageTemplate<P> {
+  frameComponent: FrameComponent<P>;
+  contentComponent: ContentComponent<P>;
+}
+
+/**
+ * To paginate ballot content, we go through the following steps:
+ *
+ * Render the content for each page:
+ * - Render the frame on the first page
+ * - Measure how much space is available inside the frame for content
+ * - Render the content for this page (given the available space) and save it
+ * - If we still have content left to render, repeat with the next page
+ *
+ * Once we have the content for each page, we render the content into the frame
+ * for each page, passing in the page number and total number of pages.
+ */
+async function paginateBallotContent<P extends Record<string, unknown>>(
+  pageTemplate: BallotPageTemplate<P>,
+  props: P,
+  document: RenderDocument
+): Promise<JSX.Element[]> {
+  const pagedContentResults: Array<PagedElementResult<P>> = [];
+
+  const { frameComponent, contentComponent } = pageTemplate;
+  do {
+    const pageFrame = frameComponent({
+      // eslint-disable-next-line vx/gts-spread-like-types
+      ...props,
+      pageNumber: pagedContentResults.length + 1,
+      totalPages: 0,
+      children: contentSlot,
+    });
+    await document.setContent('body', pageFrame);
+    const [contentSlotElement] = await document.inspectElements(
```

Comment:
Yes, agreed. I started with a stateless API like:

```ts
measureElement(element: JSX.Element): Measurements
```

I landed here because I didn't want to hide the fact that you were mutating the single shared DOM when doing this. And I wanted to use a single browser page to keep things fast.

Do you imagine creating two separate browser pages - one for scratchpad, one for final output? Or moving the scratchpad into a separate area of the same page? Or something else? Maybe creating two pages in parallel upfront could be faster than creating them on demand.

Reply:
Yeah, creating two upfront could work. Or creating one to use as a scratchpad that then gets switched into "final render" mode with a one-way toggle which would make the `measureElements` method throw? 🤷

Reply:
Yeah, that makes sense. I like the idea of adding at least some basic safety measures.